src/main/java/com/wepay/zktools/clustermgr/internal/DynamicPartitionAssignmentPolicy.java

```diff
@@ -57,7 +57,10 @@ public PartitionAssignment update(int cversion,
             int minNumPartitionsPerServer = numPartitionsToAssign / numServers;
             int remainingPartitions = numPartitionsToAssign % numServers;
 
-            // Inherit the currently assigned preferred partitions as many as possible
+            // Satisfy all preferred partition assignment, even if one or more servers ask for more than
+            // (numPartitions / numServers + 1) number of preferred partitions.
+            // But note that the assignment runs sequentially from server 1 to server N, if multiple servers claim the
+            // same partition, it will get assigned only to the first server.
             for (Map.Entry<Integer, ServerDescriptor> serverEntry : serverDescriptors.entrySet()) {
                 List<PartitionInfo> partitions = new ArrayList<>(minNumPartitionsPerServer + (remainingPartitions > 0 ? 1 : 0));
```

Comment:
The list `partitions` was initialized with a size given the previous assumption that the partition assignment will be balanced between all servers, but since this PR breaks that assumption and makes it possible that one or more servers are assigned more than `minNumPartitionsPerServer + 1` partitions, array resize will happen once for those partitions exceeding the initial capacity.
But given that we normally don't have a huge total number of partitions, I think the array resizing should not cause noticeable performance issue, so I kept the list initialization here unchanged. Other ways to go about it, either we don't set an initial capacity at all or set a larger initial capacity, don't seem to be better than the current approach. 

Let me know if you think we should change how we handle this.

Reply:
Agree, makes sense to initialize List with minimum capacity. Worst case scenario, we will end up doing array resize twice but not more than that. So, the current implementation works.